### PR TITLE
Add sign plugin task

### PR DESCRIPTION
**Pull request by satamas**
Some time ago we have added a [signing mechanism](https://blog.jetbrains.com/platform/2020/09/plugin-signing-in-marketplace/) for JetBrains plugins. Currently, all TeamCity plugins are signed by the marketplace and we are adding signature checks to the product. 
Those changes will allow plugin developers to sign their plugins using personal certificates. It is needed to safely distribute non-marketplace plugins and later it will be used for the plugin upload.

### CODE REVIEW

File: src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    List<X509Certificate> loadCertificateChain(File file) {
+        return CertificateUtils.loadCertificatesFromFile(file)
+    }
+
+    PrivateKey loadPrivateKey(File file, String password = null) {
+        return PrivateKeyUtils.loadPrivateKey(file, password?.toCharArray())
+    }
+}
```
Comment (rodm) [inline src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy:46]: @satamas the 2 methods loadCertificateChain and loadPrivateKey don't appear to be used anywhere?

File: build.gradle [hunk 2]
```diff
@@ -63,6 +63,25 @@ test {
...
+        }
+        teamcityCommonPlugin {
+            id = 'com.github.rodm.teamcity-common'
+            implementationClass = 'com.github.rodm.teamcity.TeamCityCommonPlugin'
+        }
+        teamcityEnvironmentsPlugin {
+            id = 'com.github.rodm.teamcity-environments'
+            implementationClass = 'com.github.rodm.teamcity.TeamCityEnvironmentsPlugin'
+        }
+    }
```
Comment (rodm) [inline build.gradle:84]: @satamas I don't understand why this change is needed based on the comment can you explain the problem it fixes?

**Comment by rodm** [comment]
Thanks for the PR. I notice there are no license header in some files, are you ok with me adding them with a generic copyright line?

File: src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy (hunk 1 above)
Comment (satamas) [reply src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy:46]: I've designed them as helper methods. Using those methods plugin developer can easily load their certificate and private key from the filesystem.

File: build.gradle (hunk 2 above)
Comment (satamas) [reply build.gradle:84]: When I was trying to test my plugin locally I faced a problem that it's impossible to load the plugin easily from the maven local. It was possible only if the plugin artifact was added to the buildScript as a dependency using full maven coordinates. With these changes, Gradle adds all required files, and a plugin from the local repo can be loaded by id in the same way as a plugin from the Gradle plugin portal.

**Comment by satamas** [comment]
Sure, you can add all required changes

File: build.gradle (hunk 2 above)
Comment (rodm) [reply build.gradle:84]: Ok I understand, the way I test the plugin is by adding the gradle plugin project using the 'includeBuild' function in a teamcity plugin project, I'll add a section on testing to the readme.

Comment (satamas) [reply build.gradle:84]: I've found the block I'm talking about: https://docs.gradle.org/current/userguide/plugins.html#sec:plugin_markers. Without my changes, there was no plugin markers in the local repo as far as I understand

File: src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy (hunk 1 above)
Comment (rodm) [reply src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy:46]: I've looked at this again and think it would be better to have the certificate file, private key file and password as File and String properties on the SignConfiguration and SignPluginTask and move the helper methods to the task, what do you think?

Also do you a set of steps to create a certifacte file and password file for testing?

Comment (satamas) [reply src/main/groovy/com/github/rodm/teamcity/SignConfiguration.groovy:46]: > I've looked at this again and think it would be better to have the certificate file, private key file and password as File and String properties on the SignConfiguration and SignPluginTask and move the helper methods to the task, what do you think?

I forgot to answer this question, but that wasn't a good idea. There was a reason why I did it the way I did - a certificate can be loaded not from a file but from a string or from some certificate storage. And it will be very important for a CI tools  where it is a problem to create a file but it's easy to pass string via environment variables

> Also do you a set of steps to create a certifacte file and password file for testing?

There is a [bouncy castle](https://www.bouncycastle.org/) library in dependencies that can be used to generate test certificates and private keys
